File: akt-golang/akt/router.go
package akt

import (
	"net/http"
	"strings"

	"fmt"
)

type router struct {
	handlers map[string]HandlerFunc
	roots    map[string]*node
}

func parsePattern(pattern string) []string {
	vs := strings.Split(pattern, "/")

	parts := make([]string, 0)
	for _, item := range vs {
		if item != "" {
			parts = append(parts, item)
			if item[0] == '*' {
				break
			}
		}
	}

	return parts
}

func newRouter() *router {
	return &router{handlers: make(map[string]HandlerFunc), roots: make(map[string]*node)}
}

func (router *router) addRoute(method string, pattern string, handler HandlerFunc) {
	parts := parsePattern(pattern)

	name := method + "-" + pattern

	_, ok := router.roots[method]

	if !ok {
		router.roots[method] = &node{}
	}

	router.roots[method].insert(pattern, parts, 0)

	router.handlers[name] = handler
}

func (router *router) getRoute(method string, path string) (*node, map[string]string) {
	searchParts := parsePattern(path)

	params := make(map[string]string)

	root, ok := router.roots[method]

	if !ok {
		return nil, nil
	}

	n := root.search(searchParts, 0)

	if n != nil {
		parts := parsePattern(n.pattern)
		for i, part := range parts {
			if part[0] == ':' {
				params[part[1:]] = searchParts[i]
			}
			if part[0] == '*' && len(part) > 1 {
				params[part[1:]] = strings.Join(searchParts[i:], "/")
				break
			}
		}
		return n, params
	}

	return nil, nil
}

func (router *router) handle(c *Context) {
	n, params := router.getRoute(c.Method, c.Path)
	fmt.Printf("%v", n)
	if n != nil {
		c.Params = params
		key := c.Method + "-" + n.pattern
		handler := router.handlers[key]
		c.handlers = append(c.handlers, handler)
	} else {
		c.String(http.StatusNotFound, "404 NOT FOUND: %s\n", c.Path)
	}
	c.Next()
}

File: akt-node/funcVersion/router.ts
import { AktContext } from "./context";
import { HandlerFunc } from "./akt";
import { TrieNodeType, insert, search } from "./trie";

export type AktRouter = {
  roots: Map<string, TrieNodeType>;
  handlers: Map<string, HandlerFunc>; // 路由的映射
  addRoute: (method: string, pattern: string, handler: HandlerFunc) => void; // 添加路由映射
  handle: (ctx: AktContext) => void; // 处理请求
  getRouter: (
    method: string,
    path: string
  ) => {
    node: TrieNodeType;
    params: Map<string, string>;
  } | null;
};

export const parsePattern = (pattern: string) => {
  const vs = pattern.split("/");
  const parts: string[] = [];
  for (let item of vs) {
    if (item !== "") {
      parts.push(item);
      if (item[0] === "*") {
        break;
      }
    }
  }

  return parts;
};

const getRouter = () => {
  const router: AktRouter = {
    roots: new Map(),
    handlers: new Map(),
    addRoute: (method: string, pattern: string, handler: HandlerFunc) => {
      const parts = parsePattern(pattern);
      if (!router.roots.get(method)) {
        const node: TrieNodeType = {
          pattern: "",
          part: "",
          children: [],
          isWild: false,
        };
        router.roots.set(method, node);
      }
      insert(router.roots.get(method), pattern, parts, 0);
      router.handlers.set(`${method}-${pattern}`, handler);
    },
    handle: async (ctx: AktContext) => {
      const { node, params } = router.getRouter(ctx.method, ctx.path);
      if (node) {
        ctx.params = params;
        const key = `${ctx.method}-${node.pattern}`;
        const handler = router.handlers.get(key);
        ctx.handlers.push(handler);
      } else {
        ctx.string(404, `404 NOT FOUND: ${ctx.path}`);
      }
      await ctx.next();
      ctx.res.end(ctx.resData)
    },
    getRouter: (method: string, path: string) => {
      const searchParts = parsePattern(path);
      const params: Map<string, string> = new Map();
      const root = router.roots.get(method);

      if (!root) {
        return null;
      }

      const n = search(root, searchParts, 0);

      if (n) {
        const parts = parsePattern(n.pattern);
        for (let i = 0; i < parts.length; i++) {
          if (parts[i][0] === ":") {
            params.set(parts[i].slice(1), searchParts[i]);
          }
          if (parts[i][0] === "*") {
            params.set(parts[i].slice(1), searchParts.slice(i).join("/"));
            break;
          }
        }
        return {
          node: n,
          params,
        };
      }

      return null;
    },
  };

  return router;
};

export default getRouter;
